### ISSUE TITLE
Families of APIs

### ISSUE THREAD

**Issue by janinamincer-daszkiewicz** [erasmus-without-paper/issues#46]
During the Infrastructure Forum meeting on 2024-09-18 I presented the draft proposal for introducing the families of APIs. I copy it here to start discussion. The questions we want to answer are:

1. Do we need  to regulate strictly the sets of APIs present in the network for a particular HEI?
2. What requirements are worth imposing?

The whole presentation is posted in the [EWP Developers Guide](https://esci-sd.atlassian.net/wiki/spaces/ITSC/pages/113999873/Infrastructure+Forum+and+technical+workshops).

**Groups of APIs** (certain APIs have special uses and have been omitted: Echo, Discovery, Registry, Monitoring)

1. General Purpose
    - Institutions API
    - Organizational Units API
2. Courses
     - Courses API
     - Simple Course Replication API
3. File
     - File API
4.  IIAs
    - Inter-institutional Agreements API 
    - Inter-institutional Agreements CNR API
    - Inter-institutional Agreements Approval API
    - Inter-institutional Agreements Approval CNR API
5. Mobility Factsheet
    - Mobility Factsheet API 
6.  Outgoing Mobilities LAs
    - Outgoing Mobility Learning Agreements API 
    - Outgoing Mobility Learning Agreements CNR API
7. Outgoing Mobilities (Nominations)
    - Outgoing Mobility API 
    - Outgoing Mobility CNR API
8. Incoming Mobilities ToRs
    - Incoming Mobilities ToR API 
    - Incoming Mobilities ToR CNR API
9. Incoming Mobilities (Nominations)
     - Incoming Mobilities API 
     - Incoming Mobilities CNR API

**Requirements (to be considered)**

1. If HEI implements an API, it must implement all endpoints of that API.
1.1. Stats included (also when specified as a separate API)
3. Each HEI must provide all APIs from the General Purpose group.
4. If HEI provides any APIs from the IIAs group, it must provide all APIs from this group and the Mobility Factsheet group.
5. HEI may provide APIs from the Mobility Factsheet group without providing APIs from the IIAs group.
6. If HEI sends a link to a file in any request, it must provide an API from the File group.
7. If HEI provides any APIs from the Outgoing Mobilities (Nominations) group or the Incoming Mobilities (Nominations) group, it must provide all APIs from both of these groups and the Mobility Factsheet group.

**Requirements (to be considered, courses part needs to wait for refactoring – if needed)**

8. If HEI provides any APIs from the Outgoing Mobilities LAs group, it must provide 
8.1. all APIs from this group (why?)
8.2. and the Courses group.
9. If HEI provides any APIs from the Incoming Mobilities ToRs group, it must provide 
9.1. all APIs from this group (why?)
9.2. and the Courses group.

**Comment by georgschermann** [comment]
We currently support the use case where Universities manage only incoming or outgoing LAs on one provider or split in/out across multiple providers.

For this we grouped incoming: 
-    omobility-la-cnr 

and outgoing:

- omobility-la-get
- omobility-la-index
- omobility-la-update

8,1 would eliminate this use case.

**Comment by janinamincer-daszkiewicz** [comment]
> We currently support the use case where Universities manage only incoming or outgoing LAs on one provider 

Yes, I know, we (MUCI) do the same. But may be the other providers do not like it (for some reasons). This is why we ask their opinion.

**Comment by lsanchezpe** [comment]
As i can see, Courses APIs are still in a "review recomended" state, and you post it as well in the second requirements consideration.

When this requirements are planned to be mandatory? 
And if the Courses API is not in a stable state for this time, i assume that won't be mandatory to implement it even 8.2 or 9.2 says it is. Is this correct?

**Comment by janinamincer-daszkiewicz** [comment]
> When this requirements are planned to be mandatory?

Under ESCI phase I, it can only get the status of recommendation.

> And if the Courses API is not in a stable state for this time, i assume that won't be mandatory to implement it even 8.2 or 9.2 says it is. Is this correct?

Not under Phase I. Anyway I think it makes sense to discuss what should be the recommended solution. Also because it will be a valuable income for the future contractor.

**Comment by demilatof** [comment]
What happens to an HEI that doesn't implement all the APIs of a given group?
Is it excluded from the catalogue, both in PROD and DEV?

**Comment by janinamincer-daszkiewicz** [comment]
We  allow HEIs to have errors in DEV although we also expect them to work on removing them.
In PROD - AUP (with SLA) will regulate the approach to software solutions which have bugs and bugs are not being removed promptly.
